File: android/src/main/java/id/sepran/flutter_android_speech_to_text/FlutterAndroidSpeechToTextPlugin.java
package id.sepran.flutter_android_speech_to_text;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.speech.RecognizerIntent;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;

/** FlutterAndroidSpeechToTextPlugin */
public class FlutterAndroidSpeechToTextPlugin implements FlutterPlugin, ActivityAware, MethodCallHandler, PluginRegistry.ActivityResultListener {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private Context context;

  private Result flutterResult;

  private Activity activity;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_android_speech_to_text");
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    activity = binding.getActivity();
    binding.addActivityResultListener(this);
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    activity = null;
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    activity = binding.getActivity();
    binding.addActivityResultListener(this);
  }

  @Override
  public void onDetachedFromActivity() {
    activity = null;
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
      return;
    }
    if (call.method.equalsIgnoreCase("getSpeechToText")) {
      flutterResult = result;
      final String language = call.argument("language");
      Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
      intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
      intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, language);
      intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1);

      if (intent.resolveActivity(context.getPackageManager()) != null) {
        activity.startActivityForResult(intent, 2342);
      } else {
        Toast.makeText(context, "Your Device Don't Support Speech Input", Toast.LENGTH_SHORT).show();
      }
      return;
    }
    result.notImplemented();
  }
  

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public boolean onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
    if (requestCode == 2342) {

      if (resultCode == Activity.RESULT_OK) {
        ArrayList<String> hasilList = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);

        try{
          flutterResult.success(hasilList.get(0));
        }catch(Exception e){}
        return true;
        //sendResults();
      } else {

      }
    }
    return false;
  }
}

File: android/src/main/kotlin/id/sepran/flutter_android_speech_to_text/FlutterAndroidSpeechToTextPlugin.kt
package id.sepran.flutter_android_speech_to_text

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.speech.RecognizerIntent
import android.widget.Toast
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

class FlutterAndroidSpeechToTextPlugin :
        FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener {
    var activity: Activity? = null

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var flutterResult: Result

    override fun onAttachedToEngine(
             flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
    ) {
        channel =
                MethodChannel(
                        flutterPluginBinding.binaryMessenger,
                        "flutter_android_speech_to_text"
                )
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method.equals("getSpeechToText", ignoreCase = true)) {
            flutterResult = result
            val language = call.argument<String>("language")
            val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
            intent.putExtra(
                RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
            )
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, language)
            intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1)
            if (intent.resolveActivity(context.packageManager) != null) {
                activity!!.startActivityForResult(intent, 2342)
            } else {
                Toast.makeText(
                    context,
                    "Your Device Don't Support Speech Input",
                    Toast.LENGTH_SHORT
                ).show()
            }
            return
        }
        result.notImplemented()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode == 2342) {
            if (resultCode == Activity.RESULT_OK) {
                val resultList = data!!.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
                try {
                    flutterResult.success(resultList!![0])
                } catch (e: Exception) {
                }
                return true
            }
        }
        return false
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
